### PR TITLE
feat: Add more db options

### LINKED ISSUE #1617: feat: Add neon || planetscale to relation database options (postgres || mysql)
### Is your feature request related to a problem? Please describe.

I think that for the cases that one has a preference between postgres or mysql we can choose which way to go.

### Describe the solution you'd like to see

Add an option within the command cli in which we can choose whether to use neon serverless || planetscale serverless

### Describe alternate solutions

Poder escoger si uno prefiere utilizar quizás postgres en cambio de mysql y de esa manera deducir si por default dejar neon || planetscale

### Additional information

_No response_

### LINKED ISSUE #1646: feat: Prisma Edge Client
### Is your feature request related to a problem? Please describe.

Prisma 5.4.0 added support for driver adapters to use serverless adapters over HTTP (and WS, for neon). This completes support for the Edge via PlanetScale using Prisma, previously only possible using Drizzle. Note that this also does NOT require Prisma's "Accelerate" service to use, from what I have read.

This could work really well in concert with https://github.com/t3-oss/create-t3-app/pull/1622, as selecting PlanetScale Serverless as an option would work nicely both on Prisma and on Drizzle.

### Describe the solution you'd like to see

Add a separate DB client (or just use the edge client) for edge environments:

```ts
// Import required dependencies
import { connect } from '@planetscale/database';
import { PrismaPlanetScale } from '@prisma/adapter-planetscale';
import { PrismaClient } from '@prisma/client/edge';
import { env } from "~/env.mjs";

// Setup global primitives
const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClient | undefined;
};

// Initialize Prisma Client with the PlanetScale serverless database driver
const connection = connect({ url: env.DATABASE_URL });
const adapter = new PrismaPlanetScale(connection);

export const db =
  globalForPrisma.prisma ??
  new PrismaClient({
    log: env.NODE_ENV === "development" ? ["query", "error", "warn"] : ["error"],
    adapter,
  });

if (env.NODE_ENV !== "production") globalForPrisma.prisma = db;
```

Required dependencies:
`prisma` and `@prisma/client`: 5.4.0 or higher (latest as of writing is 5.5.2)
`@prisma/adapter-planetscale`: Same version as Prisma
`@planetscale/database`: Same version as used in current deployments with PlanetScale.

This also requires the `driverAdapters` feature flag to be enabled. Since it is technically not generally availaible, this combination of options might be best marked as "experimental" in the CLI.

I also am not entirely sure about whether the edge client requires the weird `global` manipulation that the normal client does, since it operates differently (and in theory, connection pooling that causes the normal one to be problematic isn't present on the edge variant), but that requires testing that I have not yet performed.

### Describe alternate solutions

N/A

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ronanru**
Closes #1617
Closes #1646

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

## Changelog

Add a database selector, if user chose an ORM.

---

## Screenshots

image

💯

### CODE REVIEW

File: cli/src/installers/drizzle.ts
Comment (juliusmarminge) [inline cli/src/installers/drizzle.ts:37]: i don't think we want neon

File: cli/src/installers/drizzle.ts
Comment (juliusmarminge) [inline cli/src/installers/drizzle.ts:135]: all this looks prone to error at the first change we make. just having them as separate files will be easier to maintain

File: cli/template/extras/src/server/db/index-drizzle/with-postgres.ts
Comment (juliusmarminge) [inline cli/template/extras/src/server/db/index-drizzle/with-postgres.ts:2]: which one is more popular here? i've personally used postgres.js (postgres) more than node-postgres (pg)

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
I like the idea here but there are stuff we need to make sure before merging this.

What's the dev story? Should we include a db container? Recommend something like [DBngin](https://dbngin.com/)? How should the developer get the database running as quickly as possible in dev, without drifting away from the production environment?

---

And as I previously commented, we are not adding neon.

**Comment by ronanru** [comment]
> What's the dev story? Should we include a db container? 
I think adding a script that creates a db container is a great idea. I will add it in a future commit.

File: cli/template/extras/src/server/db/index-drizzle/with-postgres.ts
Comment (ronanru) [reply cli/template/extras/src/server/db/index-drizzle/with-postgres.ts]: ps is more popular, but postgres.js is smaller and faster, I will switch to it here.

**Comment by ronanru** [comment]
> Hi @ronanru , super cool idea! Do you need help on this?
> 
> FYI `neon` added a CLI, which should enable programatically creating a project/database/URL.

Thank you. This PR is done and I don't need help. 

I removed neon from this PR, Julius said Planetscale is the only DB provider T3 maintainers recommend, and they can't add stuff they don't recommend to create-t3-app.

**Comment by juliusmarminge** [comment]
Sorry for the delayed review, I'll try looking over this more closely today. We might want to add #1646 as a Prisma + PlanetScale option?

File: cli/src/installers/envVars.ts
Comment (juliusmarminge) [inline cli/src/installers/envVars.ts:89]: this seems a bit aggressive to generate a password on their behalf?

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
- Drizzle Schemas should probably have foreign keys for all configurations except PlanetScale
- Unsure about the db container. I know I raised this issue but I don't know what the best solution is. I can only talk from personal experience but I've stopped using docker for dbs nowadays and I think everyone has their own preference. We should make sure to document this in the first step and perhaps list some options 🤔

**Comment by ronanru** [comment]
> Unsure about the db container. I know I raised this issue but I don't know what the best solution is. I can only talk from personal experience but I've stopped using docker for dbs nowadays and I think everyone has their own preference. We should make sure to document this in the first step and perhaps list some options 🤔

Docker is the fastest way to start a dev database I know that works on all major operating systems. Also, it's just one small file, and you can just delete it if you don't need it. I added a paragraph to the docs about it and mentioned DBngin as an alternative.

File: cli/template/extras/prisma/schema/base-planetscale.prisma
Comment (Wundero) [inline cli/template/extras/prisma/schema/base-planetscale.prisma:10]: Should this not be `mysql` since this is a planetscale prisma schema?

Also probably needs `relationMode = "prisma"` in this section too.

File: cli/template/extras/prisma/schema/with-auth-planetscale.prisma
Comment (Wundero) [inline cli/template/extras/prisma/schema/with-auth-planetscale.prisma:15]: Likewise.

File: cli/template/extras/src/server/db/db-prisma-planetscale.ts
Comment (Wundero) [inline cli/template/extras/src/server/db/db-prisma-planetscale.ts:3]: Currently, the edge client doesn't support custom adapters (see https://github.com/prisma/prisma/issues/21394), so this would need to be the normal prisma client until that issue is resolved. Once resolved, this can change (and likely also remove some of the global fluff since connection pooling is technically unnecessary when using the edge stuff).

Comment (ronanru) [reply cli/template/extras/src/server/db/db-prisma-planetscale.ts]: Didn't know that, thank you!

**Comment by juliusmarminge** [comment]
Published a beta: 

```sh
bun create t3-app@7.25.0-beta.1
```

**Comment by juliusmarminge** [comment]
> Also, it's just one small file, and you can just delete it if you don't need it

This file is not windows compatible though right?

**Comment by ronanru** [comment]
> > Also, it's just one small file, and you can just delete it if you don't need it
> 
> This file is not windows compatible though right?

It works with WSL. I added the instructions on how to run it on Windows.

**Comment by juliusmarminge** [comment]
So sorry for the delay, but today I tested a lot of this out and I have to say awesome work!

However, using Bun, Drizzle and SQLite doesn't quite work, also Bun has it's own sqlite implementation which supposedly is faster etc. Should we use that? https://orm.drizzle.team/docs/get-started-sqlite#bun-sqlite

CleanShot 2024-01-04 at 16 26 44@2x

File: cli/template/extras/src/server/auth-app/with-drizzle.ts
Comment (juliusmarminge) [inline cli/template/extras/src/server/auth-app/with-drizzle.ts:50]: ref #1716

**Comment by juliusmarminge** [comment]
the DATABASE_URL for prisma+planetscale also doesn't work, you need to change the ssl query param:

```diff
- DATABASE_URL='mysql://user:pw@aws.connect.psdb.cloud/db?ssl={"rejectUnauthorized":true}'
+ DATABASE_URL='mysql://user:pw@aws.connect.psdb.cloud/db?sslaccept=strict' 
```

**Comment by juliusmarminge** [comment]
Current published version is `7.26.0-beta.4`

**Comment by juliusmarminge** [comment]
> So sorry for the delay, but today I tested a lot of this out and I have to say awesome work!
> 
> However, using Bun, Drizzle and SQLite doesn't quite work, also Bun has it's own sqlite implementation which supposedly is faster etc. Should we use that? https://orm.drizzle.team/docs/get-started-sqlite#bun-sqlite
> 
> EDIT: hmm doesn't look like db push works for that :/
> 
> CleanShot 2024-01-04 at 16 26 44@2x

I'm fine just not suppporting that combo to begin with... Gonna merge this

**Review by juliusmarminge — APPROVED** [review]

**Pull request merged**
